Graph workload in Jax

znado opened the pull request:
Reaches 0.24 mAP in 1671 seconds on a 96 core CPU, 8 V100 system with a per-GPU batch size of 256 (for 2048 total batch size). Have not tried to further tune the baseline, or see if we could get an even higher mAP by training longer.

Originally built off the excellent work by runame@.

Also adding a bit of formatting config that is nicer, and moving `submission_runner.py` back into the root dir for easier python importing.

runame commented on algorithmic_efficiency/workloads/ogb/ogb_jax/input_pipeline.py:
```diff
@@ -0,0 +1,161 @@
+  if not num_shards:
+    num_shards = jax.device_count()
+
+  # We will construct num_shards smaller batches and then put them together.
+  batch_size /= num_shards
```
It looks like `batch_size` is here assumed to be the global batch size and not the per-device one?

znado replied:
So given that we accumulate batches until we hit num_shards batches (`if count == num_shards:`), we should have been running with a per-device batch size of 256, for a total of 2048.

runame replied:
Ah I see now, thanks!

runame commented on algorithmic_efficiency/workloads/ogb/ogb_jax/workload.py:
```diff
@@ -0,0 +1,211 @@
+      mask_batch: Optional[spec.Tensor]) -> spec.Tensor:  # differentiable
+    loss = self._binary_cross_entropy_with_mask(
+        labels=label_batch, logits=logits_batch, mask=mask_batch)
+    mean_loss = jnp.sum(jnp.where(mask_batch, loss, 0)) / jnp.sum(mask_batch)
```
Shouldn't the loss function return the per-example instead of the mean loss?

znado replied:
Yep! Thanks for catching that. Done.

runame commented on algorithmic_efficiency/workloads/ogb/ogb_jax/workload.py:
```diff
@@ -0,0 +1,211 @@
+    return mean_loss
+
+  def _eval_metric(self, labels, logits, masks):
+    loss = self.loss_fn(labels, logits, masks)
```
This returns the mean loss but I think it should be per-example loss.

runame commented on algorithmic_efficiency/workloads/ogb/ogb_jax/workload.py:
```diff
@@ -0,0 +1,211 @@
+    eval_per_core_batch_size = 1024
+    if self._eval_iterator is None:
+      self._eval_iterator = self._build_iterator(
+          data_rng, 'validation', data_dir, batch_size=eval_per_core_batch_size)
```
Do we want to use the `validation` or `test` split here? For the validation split the mAP is >25 in the Flax example (25.2 without and 25.7 with early stopping).

znado replied:
We should probably run on both, and append `split/` to make the keys `split/metric_name`. I'll do that in a follow-up CL for all workloads.

runame commented on baselines/ogb/ogb_jax/submission.py:
```diff
@@ -0,0 +1,118 @@
+        rng,
+        update_batch_norm=True)
+    loss = workload.loss_fn(label_batch, logits_batch, mask_batch)
+    mean_loss = jnp.sum(jnp.where(mask_batch, loss, 0)) / jnp.sum(mask_batch)
```
In the current implementation `workload.loss_fn()` seems to already return the mean loss.

znado replied:
Fixed.

runame reviewed (commented):
Looks good to me! Thanks for fixing the pmap issue.

I only have questions regarding the `batch_size` and `workload.loss_fn()`, see my comments below.

znado commented on baselines/ogb/ogb_jax/submission.py:
```diff
@@ -0,0 +1,118 @@
+
+
+def get_batch_size(workload_name):
+  # Return the per-device batch size.
```
Ah good point, I like that convention more anyways. Changed.

znado commented:
Thanks so much for the thorough review!

I renamed OGB to OGBG like the dataset, fixed and documented returning the global batch size, and return per-example losses in OGBG.

runame reviewed (approved):
Thanks for addressing the comments! I have noticed that in some workloads the batch size is also just the per-device batch size, e.g. in the Jax WMT workload, see [this line](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/wmt/wmt_jax/workload.py#L291). But I will just fix this in the PyTorch WMT PR.
